Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -5,182 +5,289 @@
-        suitableSmoothNumbers = removeRows smoothNumbers
-        matrix
-          -- Here @+1@ takes into account of the further constraint given
-          -- by the negative numbers.
-          | numberOfPrimes + 1 < length mat = suitableSmoothNumbers
-          -- If the counter is odd the sieve moves one block to the right.
-          | odd counter                     = goSieving smoothNumbers (newStartingPoint + intToInteger (counter * t)) (counter + 1)
-          -- If the counter is even the sieve moves one block to the left.
-          | otherwise                       = goSieving smoothNumbers (newStartingPoint - intToInteger (counter * t)) (counter + 1)
+    goSieving :: [(Integer, I.IntMap Int)] -> [(Prime Integer, Word)] -> [(Integer, I.IntMap Int)]
+    goSieving previousDiffSmoothNumbers decompositionOfA = goSelfInitSieving previousDiffSmoothNumbers $ zip valuesOfB valuesOfC
+      where
+        a = factorBack decompositionOfA
+        valuesOfB = trace ("Ratio: " ++ show ((fromInteger (a * fromIntegral m) :: Double) / fromInteger (integerSquareRoot (2*n)))) $ filter (<= a `div` 2) $ sqrtsModFactorisation n decompositionOfA
+        valuesOfC = map (\x -> (x * x - n) `div` a) valuesOfB
+        squareRoots = map (sqrtsModPrime n . fromJust . toPrimeIntegral) factorBase
+
+        goSelfInitSieving :: [(Integer, I.IntMap Int)] -> [(Integer, Integer)] -> [(Integer, I.IntMap Int)]
+        goSelfInitSieving previousSmoothNumbers [] = goSieving previousSmoothNumbers nextDecompositionOfA
           where
-            numberOfPrimes = PS.size (foldr (\col acc -> acc <> primeSet col) mempty mat)
-            mat = fmap snd suitableSmoothNumbers
-      pure matrix
+            -- This line has to be changed
+            nextDecompositionOfA = decompositionOfA
+
+        goSelfInitSieving previousSmoothNumbers ((b, c) : otherCoeffs) = runST $ do
+          let
+            f x = a * x * x + 2 * b * x + c
+            (sievingInterval, sievingLogInterval) = V.unzip $ generateLogInterval f m
+          sievingLogIntervalM <- V.thaw sievingLogInterval
+          smoothLogSieveM sievingLogIntervalM (zip factorBase squareRoots) a b c m
+          sievedLogInterval <- V.unsafeFreeze sievingLogIntervalM
+          let
+            newSmoothNumbers = V.toList $ findLogSmoothNumbers factorBase m h decompositionOfA b $ V.zip sievingInterval sievedLogInterval
+            smoothNumbers = SS.toList . SS.fromList $ previousSmoothNumbers ++ newSmoothNumbers
+            matrixSmoothNumbers
+              -- Minimise length of matrix
+              | numberOfConstraints < length mat = trace ("Matrix dimension: " ++ show (numberOfConstraints, length mat)) $ take (numberOfConstraints + 10) smoothNumbers
+              | otherwise                        = trace ("Matrix dimension: " ++ show (numberOfConstraints, length mat)) $ goSelfInitSieving smoothNumbers otherCoeffs
```

Comment:
You can factor out tracing:
```haskell
matrixSmoothNumbers 
  | trace ... False = undefined 
  | numberOfContraints < length mat = ... 
  | otherwise = ...
```